src/sync/tracks.nim

```diff
@@ -22,73 +14,70 @@ type
   TrackExercise* = object
     slug*: string
     tests*: TrackExerciseTests
-    repoExercise: TrackRepoExercise
-
-func configJsonFile(repo: TrackRepo): string =
-  repo.dir / "config.json"
-
-func exercisesDir(repo: TrackRepo): string =
-  repo.dir / "exercises"
 
-func practiceExerciseDir(repo: TrackRepo, exercise: ConfigJsonExercise): string =
-  repo.exercisesDir / "practice" / exercise.slug
+proc `/`(head: TrackDir, tail: string): string {.borrow.}
+proc `/`(head: PracticeExercisePath, tail: string): string {.borrow.}
+proc extractFilename(exercisePath: PracticeExercisePath): string {.borrow.}
 
-func slug(exercise: TrackRepoExercise): string =
-  extractFilename(exercise.dir)
+func slug(exercisePath: PracticeExercisePath): string =
+  extractFilename(exercisePath)
 
-func testsFile(exercise: TrackRepoExercise): string =
-  exercise.dir / ".meta" / "tests.toml"
+func testsFile(exercisePath: PracticeExercisePath): string =
+  exercisePath / ".meta" / "tests.toml"
 
 func testsFile*(exercise: TrackExercise): string =
-  exercise.repoExercise.testsFile
-
-proc parseConfigJson(filePath: string): ConfigJson =
-  let json = json.parseFile(filePath)["exercises"]
-  to(json, ConfigJson)
-
-func initTrackRepoExercise(repo: TrackRepo,
-    exercise: ConfigJsonExercise): TrackRepoExercise =
-  TrackRepoExercise(dir: repo.practiceExerciseDir(exercise))
-
-proc exercises(repo: TrackRepo): seq[TrackRepoExercise] =
-  let config = parseConfigJson(repo.configJsonFile)
-
-  for exercise in config.practice:
-    result.add(initTrackRepoExercise(repo, exercise))
-
-proc initTrackExerciseTests(exercise: TrackRepoExercise): TrackExerciseTests =
-  if not fileExists(exercise.testsFile):
-    return
-
-  let tests = parsetoml.parseFile(exercise.testsFile)
-
-  for uuid, val in tests.getTable():
-    if val.hasKey("include"):
-      if val["include"].kind == Bool:
-        let isIncluded = val["include"].getBool()
-        if isIncluded:
-          result.included.incl(uuid)
+  PracticeExercisePath("").testsFile()
+
+proc getPracticeExercisePaths(trackDir: TrackDir): seq[PracticeExercisePath] =
+  let config = json.parseFile(trackDir / "config.json")["exercises"]
```

Comment:
We could consider exiting with an error if the `exercises` key does not exist, although that is extremely unlikely. It will probably only be the case for new tracks, but something to consider for later nevertheless.

Reply:
I thought the same. I'll propose improving the error for a missing `config.json` file too.

Coming soon.